sql/core/src/main/scala/org/apache/spark/sql/execution/SparkOptimizer.scala

```diff
@@ -84,6 +84,11 @@ class SparkOptimizer(
       PushPredicateThroughNonJoin,
       PushProjectionThroughLimit,
       RemoveNoopOperators) :+
+    Batch("Infer window group limit", Once,
+      InferWindowGroupLimit,
+      LimitPushDown,
+      LimitPushDownThroughWindow,
+      EliminateLimits) :+
```

Comment:
May I ask why this is different from #40142 ? At that time, we only remove one rule. So, I guess we will recover only `InferWindowGroupLimit` here.
```
- Batch("Insert window group limit", Once, InsertWindowGroupLimit) :+
```

Reply:
https://github.com/apache/spark/pull/40142 just extracts the `Limit` from `Window`. But we still need other rules about `Limit` for further optimization. e.g. `LimitPushDownThroughWindow`.